Revamp with 5-args mul! (#180)
JuliaSmoothOptimizers/LinearOperators.jl: docs/src/tutorial.md

```
In this example, the Cholesky factor is computed only once and can be used many times transparently.

## mul!

It is often useful to reuse the memory used by the operator. 
For that use, we can use `mul!` as if we were using matrices.

> We may wait to reuse the vectors preallocated by the operator in order to save memory

Reviewer: Wait?

```@example ex2
using LinearOperators, LinearAlgebra # hide
m, n = 50, 30
A = rand(m, n) + im * rand(m, n)
op = LinearOperator(A)
v = rand(n)

> al = @allocated mul!(op.Mv, op, v) # op * v, store result in op.Mv

Reviewer: Is this the intended use? It seems a bit odd to ask the user to access a temporary array stored inside `op`. Maybe the example should preallocate a new vector for storage?

Author: The user could also use a new vector, but you're right I will add more documentation about this

println("Allocation of LinearOperator mul! product = $al")
v = rand(n)
α, β = 2.0, 3.0
al = @allocated mul!(op.Mv, op, v, α, β) # α * op * v + β * op.Mv, store result in op.Mv 
println("Allocation of LinearOperator mul! product = $al")
```

You can also provide the preallocated vectors.
```@example ex2
Mv  = Array{ComplexF64}(undef, m)
Mtu = Array{ComplexF64}(undef, n)
Maw = Array{ComplexF64}(undef, n)
op  = LinearOperator(Mv, Mtu, Maw, A)
```

## Using functions

Operators may be defined from functions. They have to be based on the 5-arguments `mul!` function.
In the example below, the transposed isn't defined, but it may be inferred from the conjugate transposed. 
Missing operations are represented as `nothing`.

```@example ex1
using FFTW
function mulfft!(res, v, α, β)
  res .= α .* fft(v) .+ β .* res
end
function mulifft!(res, w, α, β)
  res .= α .* ifft(w) .+ β .* res
end
dft = LinearOperator(10, 10, false, false,

>                      (res, v, α, β) -> mulfft!(res, v, α, β),

Reviewer: Can’t we just pass `mulfft!` here?

```
In this example, the Cholesky factor is computed only once and can be used many times transparently.

## mul!

It is often useful to reuse the memory used by the operator. 

> For that use, we can use `mul!` as if we were using matrices.

Reviewer: ```suggestion
For that reason, we can use `mul!` on operators as if they were matrices.
```

> We may want to reuse the vectors preallocated by the operator in order to save memory

Reviewer: ```suggestion
One way is to use preallocated vectors:
```

```@example ex2
using LinearOperators, LinearAlgebra # hide
m, n = 50, 30
A = rand(m, n) + im * rand(m, n)
op = LinearOperator(A)
v = rand(n)
res = zeros(eltype(A), m)
al = @allocated mul!(res, op, v) # op * v, store result in res
println("Allocation of LinearOperator mul! product = $al")
v = rand(n)
α, β = 2.0, 3.0
al = @allocated mul!(res, op, v, α, β) # α * op * v + β * res, store result in res
println("Allocation of LinearOperator mul! product = $al")
```

> You can also provide the preallocated vectors, and reuse them to store the result of the product.

Reviewer: ```suggestion
You can also provide preallocated vectors, and reuse them to store operator-vector products:
```

```@example ex2
Mv  = Array{ComplexF64}(undef, m)
Mtu = Array{ComplexF64}(undef, n)
Maw = Array{ComplexF64}(undef, n)
op  = LinearOperator(Mv, Mtu, Maw, A)

> mul!(op.Mv, op, v);

Reviewer: semicolon intentional?

```
In this example, the Cholesky factor is computed only once and can be used many times transparently.

## mul!

It is often useful to reuse the memory used by the operator. 
For that reason, we can use `mul!` on operators as if we were using matrices
using preallocated vectors:

```@example ex2
using LinearOperators, LinearAlgebra # hide
m, n = 50, 30
A = rand(m, n) + im * rand(m, n)
op = LinearOperator(A)
v = rand(n)
res = zeros(eltype(A), m)
res2 = copy(res)
mul!(res2, op, v) # compile 3-args mul!
al = @allocated mul!(res, op, v) # op * v, store result in res
println("Allocation of LinearOperator mul! product = $al")
v = rand(n)
α, β = 2.0, 3.0
mul!(res2, op, v, α, β) # compile 5-args mul!
al = @allocated mul!(res, op, v, α, β) # α * op * v + β * res, store result in res

> println("Allocation of LinearOperator mul! product = $al")

Author: It was because the `res` value modifies the result of the 5-args `mul!`, so I wanted to keep `res` the same when calling the 5-args `mul!` for the first time. Should I remove it?

```

## Using functions

Operators may be defined from functions. They have to be based on the 5-arguments `mul!` function.
In the example below, the transposed isn't defined, but it may be inferred from the conjugate transposed. 
Missing operations are represented as `nothing`.
You will have deal with cases where `β == 0` and `β != 0` separately because `*` will allocate a `res` vector that
may contain `NaN` values, and `0 * NaN == NaN`.

```@example ex1
using FFTW
function mulfft!(res, v, α, β::T) where T

>   if β == zero(T)

Author: I thought it would be faster, but I can remove it and just use `β == 0` if you prefer

    res .= α .* fft(v)
  else
    res .= α .* fft(v) .+ β .* res

>   end

Author: I'm not sure but this would loop across `res ` 2 times no? What about this:
`res .= β == zero(T) ? α .* fft(v) : α .* fft(v) .+ β .* res` ?

nor hermitian will have element type `Complex128`.
This behavior may be overridden by specifying the type explicitly, e.g.,
```@example ex1
function customfunc(res, v, α, β)

>   res .= [v[1] + v[2]; v[2]] .+ β .* res

Reviewer: Isn't this a 2x2 operator? Also it looks like α is missing. We could write
```
function customfunc!(res, v, α, β)
  res[1] = (v[1] + v[2]) * α + res[1] * β
  res[2] = v[2] * α + res[2] * β
end
```
